cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,446 @@
+              .get(annotationDetails.getAnnotation().getEntities().get(0)) + 1); // + 1 because 0 specify "this"
+
+            // TODO Similar to AuthorizationEnforcer AuthenticationContext field should be generated through class
+            // rewrite. Support this too.
+            // push the authentication context
+            // this.authenticationContext
+            loadThis();
+            getField(classType, "authenticationContext", AUTHENTICATION_CONTEXT_TYPE);
+
+            // push all the actions on to the stack
+            List<Type> actionEnumSetParamTypes = new ArrayList<>();
+            for (Action action : annotationDetails.getAnnotation().getActions()) {
+              getStatic(ACTION_TYPE, action.name().toUpperCase(), ACTION_TYPE);
+              // store the Type of Enum for all the action pushed on stack as it will later be used to generate a
+              // method call instruction
+              actionEnumSetParamTypes.add(Type.getType(Enum.class));
+            }
+
+            // create a EnumSet from the above actions
+            invokeStatic(Type.getType(EnumSet.class),
+                         new Method("of", Type.getMethodDescriptor(Type.getType(EnumSet.class),
+                                                                   actionEnumSetParamTypes.toArray(
+                                                                     new Type[actionEnumSetParamTypes.size()]))));
+
+            // generate a call to AuthEnforceUtil#enforce with the above parameters on the stack
+            invokeStatic(AUTH_ENFORCE_UTIL_TYPE,
+                         new Method("enforce", Type.getMethodDescriptor(Type.VOID_TYPE,
+                                                                        AUTHORIZATION_ENFORCER_TYPE,
+                                                                        Type.getType(EntityId.class),
+                                                                        AUTHENTICATION_CONTEXT_TYPE,
+                                                                        Type.getType(Set.class))));
+          }
+        };
+      }
+      return mv;
+    }
+  }
+
+  /**
+   * Class to store all the information specified in {@link AuthEnforce} on a method
+   */
+  private class AuthAnnotation {
+    final List<String> entities;
+    final Type enforceOn;
+    final Set<Action> actions;
+
+    AuthAnnotation(List<String> entities, Type enforceOn, Set<Action> actions) {
+      this.entities = entities;
+      this.enforceOn = enforceOn;
+      this.actions = actions;
+    }
+
+    List<String> getEntities() {
+      return entities;
+    }
+
+    Type getEnforceOn() {
+      return enforceOn;
+    }
+
+    Set<Action> getActions() {
+      return actions;
+    }
+  }
+
+  /**
+   * A wrapper to store all the annotation details of method which has {@link AuthEnforce} annotation and parameters
+   * marked with {@link Name} annotation specifying the entities
+   */
+  private class AnnotationDetails {
```

Comment:
You can combine this and the `AuthAnnotation` class, right? Also, make it final static class.